hbase-server/src/main/java/org/apache/hadoop/hbase/compactionserver/CompactionThreadManager.java

```diff
@@ -19,41 +18,446 @@
+      if (completeCompactionResponse.getSuccess()) {
+        status.markComplete("Report to RS succeeded and RS accepted");
+        // move selected files to compacted files
+        compactionFilesCache.addCompactedFiles(regionInfo, cfd, selectedFileNames);
+        LOG.info("Compaction manager request complete compaction success. {}", task);
+      } else {
+        //TODO: maybe region is move, we need get latest regionserver name and retry
+        status.abort("Report to RS succeeded but RS denied");
+        LOG.warn("Compaction manager request complete compaction fail. {}", task);
+      }
+      return true;
+    } catch (IOException e) {
+      //TODO: rpc call broken, add retry
+      status.abort("Report to RS failed");
+      LOG.error("Compaction manager request complete compaction error. {}", task, e);
+      return false;
     }
-    return admin;
   }
 
-  public void requestCompaction() {
+  private List<HStoreFile> getExcludedStoreFiles(HStore store, Set<String> excludeFileNames) {
+    Collection<HStoreFile> storefiles = store.getStorefiles();
+    List<HStoreFile> storeFiles = new ArrayList<>();
+    for (HStoreFile storefile : storefiles) {
+      String name = storefile.getPath().getName();
+      if (excludeFileNames.contains(name)) {
+        storeFiles.add(storefile);
+      }
+    }
+    return storeFiles;
+  }
+
+  private HStore getStore(final Configuration conf, final FileSystem fs, final Path rootDir,
+      final TableDescriptor htd, final RegionInfo hri, final String familyName) throws IOException {
+    HRegionFileSystem regionFs = new HRegionFileSystem(conf, fs,
+        CommonFSUtils.getTableDir(rootDir, htd.getTableName()), hri);
+    HRegion region = new HRegion(regionFs, null, conf, htd, null);
+    HStore store = new HStore(region, htd.getColumnFamily(Bytes.toBytes(familyName)), conf, false);
+    OptionalLong maxSequenceId = store.getMaxSequenceId();
+    LOG.info("store max sequence id: {}", maxSequenceId.orElse(0));
+    region.getMVCC().advanceTo(maxSequenceId.orElse(0));
+    return store;
+  }
+
+  private AsyncRegionServerAdmin getRsAdmin(final ServerName sn) {
+    return server.getAsyncClusterConnection().getRegionServerAdmin(sn);
+  }
+
+  ConcurrentHashMap<String, CompactionTask> getRunningCompactionTasks() {
+    return runningCompactionTasks;
+  }
+
+  void waitForStop() {
+    compactThreadControl.waitForStop();
+  }
+
+  private void executeCompaction(CompactionTask compactionTask) {
+    try {
+      String taskName = compactionTask.getRsServerName() + "-"
+          + compactionTask.getRegionInfo().getRegionNameAsString() + "-"
+          + compactionTask.getCfd().getNameAsString() + "-" + System.currentTimeMillis();
+      compactionTask.setTaskName(taskName);
+      runningCompactionTasks.put(compactionTask.getTaskName(), compactionTask);
+      doCompaction(compactionTask);
+    } catch (Throwable e) {
+      LOG.error("Execute compaction task error: {}", compactionTask, e);
```

Comment:
the Throwable exception most likely throw by `doCompaction`, so add catch IOException and retry in `doCompaction` method. The logic here is not change?

Reply:
refer to the exception handle in Regionserver, We need increase compaction failed count and checkfilesystem is ok

Reply:
Compaction server throw ServiceException exception, and compactionOffloadManager(the master) also catch and throw ServiceException. Thus, Regionserver can catch ServiceException and do local compaction.

Reply:
I think this exception is thrown at compaction server side? How could region server know this failure?

Reply:
> Compaction server throw ServiceException exception, and compactionOffloadManager(the master) also catch and throw ServiceException. Thus, Regionserver can catch ServiceException and do local compaction.

oh, not here, I mean in the `CompactionThreadManager#requestCompaction`, added throw exception logic, If throw exception during selectfile, the ServiceException will ultimately pass to regionserver. Here run in thread pool and do compact stage, just  increase compaction failed count and checkfilesystem.

Reply:
We will retry later? How could we know the compaction task fails outside the thread pool?

Reply:
We can re-enqueue this task after puase.  No need throw exception outside the pool? I wonder whether we need retry here, If filesystem is not available, we will shutdown the compactionserver

Reply:
OK, I will add retry, But the requestCompaction has responded, not need to throw exception outside. If retry finally failed, just increase compaction failed count and checkfilesystem.

Reply:
I still do not get the whole picture here... Could you please explain how to deal with compaction failure?

In the old implementation, the region server will mark the compaction as failure and schedule a new one later. But now, the region server just call requestCompaction to master, and master will delegate the compaction to compaction server, then how can we deal with the final failure on compaction server?

Reply:
We invoke compaction in these situations: rsRpcService#compactRegion(user trigger), CompactionChecker, loadRecoveredHFiles. The `CompactionChecker` swallow the exception compaction throw and `rsRpcService` ,`loadRecoveredHFiles` rethrow exception. They do not have retry logic, `CompactionChecker` can trigger another compaction after period. User can deal with the ServiceException thrown by `rsRpcService#compactRegion`, and decide to retry or not.

Reply:
now, the logic of regionserver and compactionserver is equivalent.
1)In regionserver, before selectfile we can throw IOException:
https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CompactSplit.java#L344
and once the file is selected and enqueue thread pool, we swallow the exception and just record the failed.
https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CompactSplit.java#L630
2) In compactionServer, we can throw exception when select file failed, and the ServiceException will ultimately pass to regionserver:
https://github.com/apache/hbase/blob/8d517e2be61c89ca15dcd32210eb08ea46046605/hbase-server/src/main/java/org/apache/hadoop/hbase/compactionserver/CompactionThreadManager.java#L146

_CsRpcservice throw ServiceException and master will catch it:_
https://github.com/apache/hbase/blob/8d517e2be61c89ca15dcd32210eb08ea46046605/hbase-server/src/main/java/org/apache/hadoop/hbase/compactionserver/CSRpcServices.java#L109

_Master throw ServiceException and regionserver will catch it:_
https://github.com/apache/hbase/blob/8d517e2be61c89ca15dcd32210eb08ea46046605/hbase-server/src/main/java/org/apache/hadoop/hbase/master/compaction/CompactionOffloadManager.java#L176

_Regionserver catch ServiceException and retrun false:_
https://github.com/apache/hbase/blob/8d517e2be61c89ca15dcd32210eb08ea46046605/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java#L3780

_Regionserver retry compaction in local_
https://github.com/apache/hbase/blob/8d517e2be61c89ca15dcd32210eb08ea46046605/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java#L1937

and once enqueue thread pool, we swallow the exception and just record the failed. 
https://github.com/apache/hbase/blob/8d517e2be61c89ca15dcd32210eb08ea46046605/hbase-server/src/main/java/org/apache/hadoop/hbase/compactionserver/CompactionThreadManager.java#L390

Reply:
The last case is what I concern here. You said

"CompactionChecker can trigger another compaction after period"

The CompactionChecker is on region server or compaction server? I guess it is on region server? If compaction server fails to comapct, then trigger another compaction is fine, but what if the compaction is still ongoing? Master need to send the compaction request to the same compaction server? Do we have this logic now?

Reply:
Yes, the CompactionChecker is on regionserver. The second compaction request will send to same compaction server, the compaction trigger by CompactionChecker also have same logic now. However if compaction server catch exception while do compact and failed to pass checkfilesystem, the compaction server will shutdown, we will chose another compactionserver

Reply:
Is this logic in this PR or in previous PR? I haven't found the related code in this PR's diff...

Reply:
previous HBASE-25968 
https://github.com/apache/hbase/pull/3378/files#diff-3df66a3c675e39368210007be9262484b0202091d0297f50a161410d39808591R153

Reply:
```
  private ServerName selectCompactionServer(CompactRequest request) throws ServiceException {
    List<ServerName> compactionServerList = getOnlineServersList();
    if (compactionServerList.size() <= 0) {
      throw new ServiceException("compaction server is not available");
    }
    // TODO: need more complex and effective way to manage compaction of region to CS mapping.
    //  maybe another assignment and balance module
    long index = (request.getRegionInfo().getStartKey().hashCode() & Integer.MAX_VALUE)
        % compactionServerList.size();
    return compactionServerList.get((int) index);
  }
```

The code is not stable. If there are compaction server crash, it will effect requests which are not on the crashed server.

Anyway, we could improve this method later.

Reply:
Yes, we can promote the logic here later.